### PR TITLE
Ocelots - Yvette W.

### PR DESCRIPTION
**Pull request by dp-wu**

### CODE REVIEW

File: swap_meet/clothing.py [hunk 1]
```diff
@@ -1,2 +1,17 @@
...
+
+    def __init__(self, id=None, fabric="Unknown", condition=0):
+        super().__init__(id, condition)
+        self.fabric = fabric
+    
+    def __str__(self):
+        """
+        return a message about the clothing instance
+        """
+        stringified = super().__str__() + f". It is made from {self.fabric} fabric."
```
Comment (kelsey-steven-ada) [inline swap_meet/clothing.py:15]: Since we're using a f-string already, another option would be to call the superclass function from a placeholder:
```python
stringified = f"{super().__str__()}. It is made from {self.fabric} fabric."
```

File: swap_meet/item.py [hunk 2]
```diff
@@ -1,2 +1,39 @@
...
+        """
+        conditions = {
+            5: "mint",
+            4: "excellent",
+            3: "very good",
+            2: "good",
+            1: "fair",
+            0: "heavily used",
+        }
+        return conditions[round(self.condition)]
```
Comment (kelsey-steven-ada) [inline swap_meet/item.py:39]: Nice use of round to handle decimal condition values!

File: swap_meet/vendor.py [hunk 3]
```diff
@@ -1,2 +1,165 @@
...
+        """
+        remove an item from inventory
+        return removed item or None if not found
+        """
+        removed = item
+        if item in self.inventory:
+            self.inventory.remove(item)
+        else:
+            removed = None
+        return removed
```
Comment (kelsey-steven-ada) [inline swap_meet/vendor.py:26]: If we default `removed` to `None`, we could simplify this block by removing the `else`:
```python
        removed = None
        if item in self.inventory:
            self.inventory.remove(item)
            removed = item

        return removed
```

File: swap_meet/vendor.py [hunk 4]
```diff
@@ -1,2 +1,165 @@
...
+    
+    def get_by_id(self, id):
+        """
+        find if item of an given id exist in inventory
+        return found item or None
+        """
+        found = None
+        for item in self.inventory:
+            if item.id == id:
+                found = item
```
Comment (kelsey-steven-ada) [inline swap_meet/vendor.py:36]: We could leave the function early once we find an item with a matching `id`. Think about a case where we might have 1,000 items and the very first has the id we're looking for. If we don't break out of the loop, we will iterate 999 more times than necessary. If you want to keep a single return, we could use a `break` statement to prevent looping over further items.

File: swap_meet/vendor.py [hunk 5]
```diff
@@ -1,2 +1,165 @@
...
+        return found
+
+    def swap_items(self, other_vendor, my_item, their_item):
+        """
+        swap given items between my inventory and other vendor
+        return True if targeted items exist and swapped, False otherwise
+        """
+        swapped = False
+        found_mine = None if my_item is None else self.get_by_id(my_item.id)
+        found_theirs = None if their_item is None else other_vendor.get_by_id(their_item.id)
```
Comment (kelsey-steven-ada) [inline swap_meet/vendor.py:46]: To make these lines a bit shorter to follow PEP8 guidelines, I would consider creating a validation check at the top of the function for the `None` checks:
```python
if my_item is None or their_item is None:
    return false

swapped = False
found_mine = self.get_by_id(my_item.id)
found_theirs = other_vendor.get_by_id(their_item.id)
```
Another option would be breaking the code over a couple lines:
```python
swapped = False
found_mine = (None if my_item is None 
    else self.get_by_id(my_item.id))
found_theirs = (None if their_item is None 
    else other_vendor.get_by_id(their_item.id))
```

File: swap_meet/vendor.py [hunk 6]
```diff
@@ -1,2 +1,165 @@
...
+        if found_mine is not None and found_theirs is not None:
+            other_vendor.add(self.remove(my_item))
+            self.add(other_vendor.remove(their_item))
```
Comment (kelsey-steven-ada) [inline swap_meet/vendor.py:49]: I like the idea of what you're doing, using the return values of the `remove` functions as the arguments to `add`, but I would consider that there's no performance benefit to nesting these. In shared code, I would suggest un-nesting the function calls to make the lines even easier to read and understand quickly.

File: swap_meet/vendor.py [hunk 7]
```diff
@@ -1,2 +1,165 @@
...
+    
+    def swap_first_item(self, other_vendor):
+        """
+        swap first item of inventory from self and other vendor
+        return True if both inventories are not empty and swapped, False otherwise
+        """
+        swapped = False
+        if self.inventory and other_vendor.inventory:
+            other_vendor.add(self.remove(self.inventory[0]))
+            self.add(other_vendor.remove(other_vendor.inventory[0]))
```
Comment (kelsey-steven-ada) [inline swap_meet/vendor.py:61]: Since the overall performance would be similar, we could reuse the `swap_items` method to reduce repeated/similar code between these functions.
```python
my_item = self.inventory[0]
their_item = other_vendor.inventory[0]
swapped = self.swap_items(other_vendor, my_item, their_item)
```

File: swap_meet/vendor.py [hunk 8]
```diff
@@ -1,2 +1,165 @@
...
+    def get_by_category(self, category):
+        """
+        return a list of item from inventory of given category
+        return empty list if no item in the inventory is of the give category
+        """
+        items = list()
+        for item in self.inventory:
+            if item.get_category() == category:
+                items.append(item)
+        return items
```
Comment (kelsey-steven-ada) [inline swap_meet/vendor.py:74]: Notice this pattern of: 
```python
result_list = []
for element in source_list:
    if some_condition(element):
        result_list.append(element)
```

can be rewritten using a list comprehension as:
```python
result_list = [element for element in source_list if some_condition(element)]
```

Which here would look like:
```python
items = [item for item in self.inventory if item.get_category() == category]
```

At first, this may seem more difficult to read, but comprehensions are a very common python style, so I encourage y’all to practice working with them!

File: swap_meet/vendor.py [hunk 9]
```diff
@@ -1,2 +1,165 @@
...
+    
+    def get_best_by_category(self, category):
+        # I was thinking implement it with only lambda function, but couldn't figure out how to :/
+        """
+        return None if inventory is empty or no such category DNE in the inventory
+        else return the first best conditioned item of the category
+        """
+        items = self.get_by_category(category)
+        item = None if not items else max(items, key=lambda i: i.condition)
```
Comment (kelsey-steven-ada) [inline swap_meet/vendor.py:83]: Great use of `max` with a `lambda`! I would suggest using a more descriptive variable name inside the lambda.

File: swap_meet/vendor.py [hunk 10]
```diff
@@ -1,2 +1,165 @@
...
+        return item
+    
+    def swap_best_by_category(self, other_vendor, my_priority, their_priority):
+        """
+        find best conditioned item of given category of each person (self vendor, other vendor)
+        if desired items are in eachother's inventory, swap them
+        if any items not exist or inventory empty, return None
+        """
+        i_have = self.get_best_by_category(their_priority)
+        vendor_has = other_vendor.get_best_by_category(my_priority)
```
Comment (kelsey-steven-ada) [inline swap_meet/vendor.py:93]: These variable names are solid, and we could be even more descriptive. Using line 92 as an example, some other options could be `their_interest_item`, `best_item_their_interest`, or `my_best_of_their_priority `

File: swap_meet/vendor.py [hunk 11]
```diff
@@ -1,2 +1,165 @@
...
+        return self.swap_items(other_vendor, i_have, vendor_has)
+    
+    def display_inventory(self, category=None):
+        """
+        print all inventory of some inventory of a given category
+        print a message if inventory is empty or no inventory of the given category
+        """
+        items = self.inventory if category is None else self.get_by_category(category)
```
Comment (kelsey-steven-ada) [inline swap_meet/vendor.py:101]: Nice conditional assignment! I don't think this was specified in the README, but if we wanted to use the full inventory if the user passes an empty string for the category, I would suggest checking `if not category`.

File: swap_meet/vendor.py [hunk 12]
```diff
@@ -1,2 +1,165 @@
...
+        result = str()
+        if items:
+            for ind, item in enumerate(items):
+                # this part took me (and Sam) forever to figure out :/
+                # and Monica told me she prints out line by line so never encountered such problem
+                result += '{}. {}'.format(ind+1, item.__str__())
```
Comment (kelsey-steven-ada) [inline swap_meet/vendor.py:107]: Because we have implemented the `__str__` method, this is one of the situations where python will implicitly call it for us, we don't need to manually invoke it:
```python
result += '{}. {}'.format(ind+1, item)
```

File: swap_meet/vendor.py [hunk 13]
```diff
@@ -1,2 +1,165 @@
...
+                result += '\n' if ind+1 < len(items) else ''
```
Comment (kelsey-steven-ada) [inline swap_meet/vendor.py:108]: The string.format() function is an older way of formatting, since f-strings were introduced to python they are the general recommendation for formatting as they tend to be much more human-readable.
```python
result += f"{i+1}. {result_list[i]}"
```

File: swap_meet/vendor.py [hunk 14]
```diff
@@ -1,2 +1,165 @@
...
+    
+    def display_inventory(self, category=None):
+        """
+        print all inventory of some inventory of a given category
+        print a message if inventory is empty or no inventory of the given category
+        """
+        items = self.inventory if category is None else self.get_by_category(category)
+        result = str()
+        if items:
+            for ind, item in enumerate(items):
```
Comment (kelsey-steven-ada) [inline swap_meet/vendor.py:104]: If we don't want to manually add the `+1` to the index, `enumerate` can take a second parameter that tells the counter `ind` where to start: `enumerate(items, start=1)`

File: swap_meet/vendor.py [hunk 15]
```diff
@@ -1,2 +1,165 @@
...
+                # this part took me (and Sam) forever to figure out :/
+                # and Monica told me she prints out line by line so never encountered such problem
```
Comment (kelsey-steven-ada) [inline swap_meet/vendor.py:106]: I'm glad you reached out to office hours and found a way to handle the edge case!

File: tests/integration_tests/test_wave_01_02_03.py [hunk 16]
```diff
@@ -2,8 +2,7 @@
 from swap_meet.vendor import Vendor
 from swap_meet.item import Item
 
-@pytest.mark.skip
-@pytest.mark.integration_test
+
```
Comment (kelsey-steven-ada) [inline tests/integration_tests/test_wave_01_02_03.py:5]: For integration tests, we want to leave the decorator `@pytest.mark.integration_test` uncommented. This decorator tells pytest that this is an integration test, so we should run it after all of the unit tests.

File: tests/unit_tests/test_wave_01.py [hunk 17]
```diff
@@ -49,7 +48,8 @@ def test_removing_not_found_returns_none():
 
     result = vendor.remove(item)
 
-    raise Exception("Complete this test according to comments below.")
+    # raise Exception("Complete this test according to comments below.")
     # *********************************************************************
     # ****** Complete Assert Portion of this test **********
     # *********************************************************************
+    assert result == None
```
Comment (kelsey-steven-ada) [inline tests/unit_tests/test_wave_01.py:55]: Nice assert, we definitely want to confirm our return value is `None`. We generally also want to confirm that there weren't side affects or changes to the input that we didn't intend. What could we check about the vendor's inventory to confirm that its content didn’t change?

File: tests/unit_tests/test_wave_03.py [hunk 18]
```diff
@@ -131,7 +131,8 @@ def test_swap_items_from_their_empty_returns_false():
 
     result = fatimah.swap_items(jolie, item_b, nobodys_item)
 
-    raise Exception("Complete this test according to comments below.")
+    # raise Exception("Complete this test according to comments below.")
     # *********************************************************************
     # ****** Complete Assert Portion of this test **********
     # *********************************************************************
+    assert result is False
```
Comment (kelsey-steven-ada) [inline tests/unit_tests/test_wave_03.py:138]: The feedback in wave 1 applies here as well, what else would be helpful to assert to ensure there were no unintended side effects?

File: tests/unit_tests/test_wave_08.py [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+    assert len(jesse.inventory) == 3
+    assert len(tai.inventory) == 3
+    assert item_a in tai.inventory
+    assert item_b in tai.inventory
+    assert item_c in tai.inventory
+    assert item_d in jesse.inventory
+    assert item_e in jesse.inventory
+    assert item_f in jesse.inventory
+
+def test_bad_instance_id():
```
Comment (kelsey-steven-ada) [inline tests/unit_tests/test_wave_08.py:77]: Love the added tests! 

For this one, I would consider adding some descriptors to the name. Since we're specifically looking at testing that we raise an error if an item subclass is instantiated with a bad id, we could use something like `test_create_item_subclass_invalid_id_raises_error` or `test_instantiation_raises_error_if_item_subclass_id_invalid`. Splitting this out into tests for each subclass could help reduce the length of the name since we could be more specific and could help us debug faster.

**Review by kelsey-steven-ada — COMMENTED** [review]
Great work Yvette! I've left a mix of suggestions and questions to consider for feedback. Please reply here on Github or reach out on Slack if there's anything I can clarify =]
